### PR TITLE
Flintlock revolver

### PR DESCRIPTION
**Pull request by misterprimus**
Spawn and use a flintlock revolver.
#### Summary

```SUMMARY: Content "Adds flintlock revolver-type weapons."```
#### Purpose of change

To make this gun craftable: https://en.wikipedia.org/wiki/Elisha_Collier
#### Describe the solution

Makes flintlock revolver-type weapons craftable. This requires a rifling button, which can now be crafted.
#### Describe alternatives you've considered

N/A.
#### Testing

Spawned and reloaded some revolvers.

### CODE REVIEW

**Comment by tenmillimaster** [comment]
- Could you please explain why one would bother with rifling on a homemade flintlock revolver?
- Why not make simpler pepperboxes?
- Manufacturing rifling buttons requires specialized steels, precision metrology, and machining/grinding. I would argue cut rifling is more feasible for the survivor.

File: data/json/recipes/weapon/ranged.json [hunk 1]
```diff
@@ -155,6 +155,72 @@
...
+  {
+    "type": "recipe",
+    "result": "longrifle_flintlock_revolver",
+    "category": "CC_WEAPON",
+    "subcategory": "CSC_WEAPON_RANGED",
+    "skill_used": "fabrication",
+    "skills_required": [ "rifle", 5 ],
+    "difficulty": 10,
+    "time": "560 m",
+    "autolearn": true,
```
Comment (LaVeyanFiend) [inline data/json/recipes/weapon/ranged.json:167]: You can't both autolearn a recipe and it explicitly be a replica of someone else's design.

File: data/json/recipes/weapon/ranged.json [hunk 2]
```diff
@@ -155,6 +155,72 @@
...
+  {
+    "type": "recipe",
+    "result": "carbine_flintlock_revolver",
+    "category": "CC_WEAPON",
+    "subcategory": "CSC_WEAPON_RANGED",
+    "skill_used": "fabrication",
+    "skills_required": [ "rifle", 5 ],
+    "difficulty": 10,
+    "time": "560 m",
+    "autolearn": true,
```
Comment (LaVeyanFiend) [inline data/json/recipes/weapon/ranged.json:189]: ^

File: data/json/recipes/weapon/ranged.json [hunk 3]
```diff
@@ -155,6 +155,72 @@
...
+  {
+    "type": "recipe",
+    "result": "flintlock_revolver",
+    "category": "CC_WEAPON",
+    "subcategory": "CSC_WEAPON_RANGED",
+    "skill_used": "fabrication",
+    "skills_required": [ "pistol", 5 ],
+    "difficulty": 10,
+    "time": "550 m",
+    "autolearn": true,
```
Comment (LaVeyanFiend) [inline data/json/recipes/weapon/ranged.json:211]: ^

**Comment by misterprimus** [comment]
> * Could you please explain why one would bother with rifling on a homemade flintlock revolver?
> * Why not make simpler pepperboxes?
> * Manufacturing rifling buttons requires specialized steels, precision metrology, and machining/grinding. I would argue cut rifling is more feasible for the survivor.

I got rid of the rifling on the revolver. Yeah, you're right, at that short a barrel, not worth it.
I wanted to add something with accuracy, homemade, and multiple rounds. This fit the bill. A pepperbox would be terribly inaccurate or, if barrels are long, impossible to lift.
A homemade rifling button can be made like this:
https://www.youtube.com/watch?v=D43ZeYu9dnM&t
Obviously I wouldn't expect a sniper rifle to be a result of using such a button, but a fairly good blackpowder rifle seems possible.

File: data/json/recipes/weapon/ranged.json (hunk 1 above)
Comment (misterprimus) [reply data/json/recipes/weapon/ranged.json]: fixed

**Comment by tenmillimaster** [comment]
That method of making button definitely looks practical. Thank you for sharing.

**Comment by misterprimus** [comment]
> That method of making button definitely looks practical. Thank you for sharing.

That's why I put the drill lvl 2 as a requirement. That's basically the "dremel." Obviously no solution will be perfect, but that has more to do with the abstraction in Cat's crafting system. The process IRL will obviously vary a lot depending on the size metal cylinder you find, etc... Just like we abstract "wrench" for pretty much all the possible wrench sizes, types, sockets, etc...

**Comment by kevingranade** [comment]
As we discussed, I find your source incredibly unconvincing.
It requires finding a piece of tool steel that's an incredibly precise size.
The resulting rifling was notably rough.
The button broke on first use, which is a huge red flag.

The in-game depiction is problematic too.
The button recipe simply requires "any steel you find".  It literally allows you to make the button and barrel from the same material, which is impossible. At best it would require harvesting from specifically tool steel sources, at worst, I'm not convinced that tool steel is going to be sufficient.
The button recipe has no indication that each button would be calibre specific.

From an overall game balance point of view, all of these gun recipes are incredibly cursory, and I'm much more inclined to back them out than to keep adding more.
There's no drilling, no material requirements, no fitting, and the time requirements are absurdly low.

**Comment by misterprimus** [comment]
> As we discussed, I find your source incredibly unconvincing.
> It requires finding a piece of tool steel that's an incredibly precise size.
> The resulting rifling was notably rough.
> The button broke on first use, which is a huge red flag.
> The in-game depiction is problematic too.
> The button recipe simply requires "any steel you find".  It literally allows you to make the button and barrel from the same material, which is impossible. At best it would require harvesting from specifically tool steel sources, at worst, I'm not convinced that tool steel is going to be sufficient.
> The button recipe has no indication that each button would be calibre specific.
> From an overall game balance point of view, all of these gun recipes are incredibly cursory, and I'm much more inclined to back them out than to keep adding more.
> There's no drilling, no material requirements, no fitting, and the time requirements are absurdly low.

1. It has to be precise only if you are designing it for a commercially available caliber. The guy in the video looks like he was trying to make a more or less modern firearm with this to shoot rigid metal cased ammo he buys online. With a flintlock, you're going to be making your own ammo and using a soft paper cartridge, so in that case you'd likely make the gun first and then make the ammo fit whatever hole you have.
2. Yes, the rifling is rough, but the gun is only as accurate as a regular flintlock.
3. If you are concerned with the button breaking, I can make the button be a component instead of a tool to simulate it being single use only due to breaking.

I agree the guns in this game (along with lots of other complex things like vehicle repairs) are problematic due to being abstracted. We mostly have just one type of steel, as opposed to the numerous grades of steel that exist IRL. I compare it to the "wrench" being just one type of wrench, which ignores the numerous possible ratchet types, sockets sizes, socket depths, etc... Same with abstract "alternators" fitting pretty much any vehicle or engine repairs being much easier in the game compared to IRL (where they would often be impossible).

Gun repairs, too, are in many cases impossible IRL due to modern guns requiring perfectly machined parts. In reality you would need to find a gun with a compatible intact part, disassemble both guns, and reassemble one that works using parts form both.

All that considered, I'd say this weapon is much more conservative than the majority of the things in this game currently and is hardly a departure from reality. Hopefully one day we have a functioning stationary machinery crafting system (with drill presses, lathes, etc...) as well as an interchangeable parts repair system with the condition of each component (barrel, trigger, etc...) being tracked individually.

**Comment by Rail-Runner** [comment]
I think it'd be best to just make them smoothbore if it turns out to be too difficult to make the barrel rifled.

**Comment by misterprimus** [comment]
> I think it'd be best to just make them smoothbore if it turns out to be too difficult to make the barrel rifled. I think they'd still be distinctive from the other flintlock weapons since they don't need to reload after each shot.
> Also as related to what you replied to Kevin, you'd still need different types of steel to pull this off. That said, it's relatively easy to make some hardened (high/medium-carbon) steel for example, it's just that adding new types of steel would need lots of infrastructure work since right now "generic" steel is used for so many recipes.

The whole "rifled" thing is mostly just semantics anyway. I doubt the game simulates long range gunfire accurately enough to even have riflings matter.

I just wanted to have a DIY multishot blackpowder solution and since this gun actually does exist and was made a long time ago I figured it would be possible to make one IRL, especially since even parts as difficult as the rifled barrel have been done IRL without any advanced machine tools.

Many of the complaints seem to be things that are simply not detailed enough in the game in general, but I don't know if that means we need a near-total crafting freeze until it's resolved as that could take months or years to work out if it ever gets done. Each engine would need engine-specific components, you'd have to have pipes of different materials, diameters, lengths, instead of our generic "pipe," guns would need specific gun parts, etc... I kind of did something like that for guns in STALKER's Dead Air mod (it was an addon I created) and I had to run the thing through excel just to come up with all the parts for the game's very limited arsenal. I can't even imagine doing that for all the car parts and weapons in this game. It would be a truly insurmountable task if we are to avoid all or most abstraction/simplification.

**Comment by Rail-Runner** [comment]
> > I think it'd be best to just make them smoothbore if it turns out to be too difficult to make the barrel rifled. I think they'd still be distinctive from the other flintlock weapons since they don't need to reload after each shot.
> > Also as related to what you replied to Kevin, you'd still need different types of steel to pull this off. That said, it's relatively easy to make some hardened (high/medium-carbon) steel for example, it's just that adding new types of steel would need lots of infrastructure work since right now "generic" steel is used for so many recipes.
> 
> The whole "rifled" thing is mostly just semantics anyway. I doubt the game simulates long range gunfire accurately enough to even have riflings matter.
> 
> I just wanted to have a DIY multishot blackpowder solution and since this gun actually does exist and was made a long time ago I figured it would be possible to make one IRL, especially since even parts as difficult as the rifled barrel have been done IRL without any advanced machine tools.
> 
> Many of the complaints seem to be things that are simply not detailed enough in the game in general, but I don't know if that means we need a near-total crafting freeze until it's resolved as that could take months or years to work out if it ever gets done. Each engine would need engine-specific components, you'd have to have pipes of different materials, diameters, lengths, instead of our generic "pipe," guns would need specific gun parts, etc... I kind of did something like that for guns in STALKER's Dead Air mod (it was an addon I created) and I had to run the thing through excel just to come up with all the parts for the game's very limited arsenal. I can't even imagine doing that for all the car parts and weapons in this game. It would be a truly insurmountable task if we are to avoid all or most abstraction/simplification.

It isn't actually making things much more difficult to have 2 or 3 kinds of steel to make things from, compared to having many pipes of different diameters or whatever. I think the ones that would matter most are mild steel (generally speaking, something you'd make vehicles from), hardened steel (something you'd want to make blades from), and high-speed steel if at some point we want to get into making machine tools, but since we don't have them yet, it could be ignored until then.

**Comment by tenmillimaster** [comment]
>That's why I put the drill lvl 2 as a requirement. That's basically the "dremel." Obviously no solution will be perfect, but that has more to do with the abstraction in Cat's crafting system. The process IRL will obviously vary a lot depending on the size metal cylinder you find, etc... Just like we abstract "wrench" for pretty much all the possible wrench sizes, types, sockets, etc...

We don't have dremels in game, for starters.
You need, as Kevin mentioned, a tool steel of a suitable sort.
 Given that each button is specific to each caliber, 1 per is best. You would steel need machinist's measurement tools, a press of some sort. And considering the very, very, very rough rifling, this isn't really expandable to anything larger than a SMG. A helical jig to cut the rifling rather than button or broach it would be better suited.

Anyways, rifling the rifles and pistols seems entirely unnecessary. 
*Practically* speaking, the priming mechanism on the colliers, the pawl, has to be done away with. It's just too fine, too complicated to make. Too many fiddly bits.

Which means the practical version of the revolver would need to be primed separately for each shot. Which requires the guns to take two ammo types. Which we don't support (yet).

**Comment by misterprimus** [comment]
> > > I think it'd be best to just make them smoothbore if it turns out to be too difficult to make the barrel rifled. I think they'd still be distinctive from the other flintlock weapons since they don't need to reload after each shot.
> > > Also as related to what you replied to Kevin, you'd still need different types of steel to pull this off. That said, it's relatively easy to make some hardened (high/medium-carbon) steel for example, it's just that adding new types of steel would need lots of infrastructure work since right now "generic" steel is used for so many recipes.
> > 
> > 
> > The whole "rifled" thing is mostly just semantics anyway. I doubt the game simulates long range gunfire accurately enough to even have riflings matter.
> > I just wanted to have a DIY multishot blackpowder solution and since this gun actually does exist and was made a long time ago I figured it would be possible to make one IRL, especially since even parts as difficult as the rifled barrel have been done IRL without any advanced machine tools.
> > Many of the complaints seem to be things that are simply not detailed enough in the game in general, but I don't know if that means we need a near-total crafting freeze until it's resolved as that could take months or years to work out if it ever gets done. Each engine would need engine-specific components, you'd have to have pipes of different materials, diameters, lengths, instead of our generic "pipe," guns would need specific gun parts, etc... I kind of did something like that for guns in STALKER's Dead Air mod (it was an addon I created) and I had to run the thing through excel just to come up with all the parts for the game's very limited arsenal. I can't even imagine doing that for all the car parts and weapons in this game. It would be a truly insurmountable task if we are to avoid all or most abstraction/simplification.
> 
> It isn't actually making things much more difficult to have 2 or 3 kinds of steel to make things from, compared to having many pipes of different diameters or whatever. I think the ones that would matter most are mild steel (generally speaking, something you'd make vehicles from), hardened steel (something you'd want to make blades from), and high-speed steel if at some point we want to get into making machine tools, but since we don't have them yet, it could be ignored until then. However, someone does have to go around and add the newly-added material to all the recipes in which it's supposed to be used.

Added metals would hardly solve most of the realism/depth issues surrounding crafting in this game. The most difficult part of crafting isn't finding the right metal, but shaping it. And even just the new material would requiring, like you said, having someone adjust each recipe.

**Comment by misterprimus** [comment]
> > That's why I put the drill lvl 2 as a requirement. That's basically the "dremel." Obviously no solution will be perfect, but that has more to do with the abstraction in Cat's crafting system. The process IRL will obviously vary a lot depending on the size metal cylinder you find, etc... Just like we abstract "wrench" for pretty much all the possible wrench sizes, types, sockets, etc...
> 
> We don't have dremels in game, for starters.
> You need, as Kevin mentioned, a tool steel of a suitable sort. (which we have, `hardsteel`, but we don't have the intermediate items you'd want to scavenge this from, like endmills, reamers, drill bits, hardened dowel pins, that sort of thing).
> We'd also want a drill press so you could turn the rod down to the appropriate size. (which we don't have, also)
> Given that each button is specific to each caliber, 1 per is best. You would steel need machinist's measurement tools, a press of some sort. And considering the very, very, very rough rifling, this isn't really expandable to anything larger than a SMG. A helical jig to cut the rifling rather than button or broach it would be better suited.
> 
> Anyways, rifling the rifles and pistols seems entirely unnecessary.
> _Practically_ speaking, the priming mechanism on the colliers, the pawl, has to be done away with. It's just too fine, too complicated to make. Too many fiddly bits.
> image
> 
> Which means the practical version of the revolver would need to be primed separately for each shot. Which requires the guns to take two ammo types. Which we don't support (yet).
> 
> > The whole "rifled" thing is mostly just semantics anyway. I doubt the game simulates long range gunfire accurately enough to even have riflings matter.
> 
> Rifling matters. A keyholing projectile will have worse armor penetration than a properly stabilized one, accuracy aside.

So what is the suggestion? Everything you suggested you followed up with "we don't have the infrastructure for that yet," which I mostly agree with, but how would that be solved?

**Comment by tenmillimaster** [comment]
> So what is the suggestion? Everything you suggested you followed up with "we don't have the infrastructure for that yet," which I mostly agree with, but how would that be solved?

By adding it.

**Comment by misterprimus** [comment]
> > So what is the suggestion? Everything you suggested you followed up with "we don't have the infrastructure for that yet," which I mostly agree with, but how would that be solved?
> 
> By adding it.

You mean the different metals, update all the recipes and items to be one type of metal or another, add powered drill presses and assorted other stationary tools, and implement multiple ammo types just to add a revolver?!

**Comment by kevingranade** [comment]
> Many of the complaints seem to be things that are simply not detailed enough in the game in general, but I don't know if that means we need a near-total crafting freeze until it's resolved as that could take months or years to work out if it ever gets done.

This is hyperbolic and missing the point, which is that the things you're trying to make craftable are in no way necessary for the game.  Crafting a black powder revolver from scratch is an utterly ludicrous idea, because there are millions of working guns just lying around in the game world. If you were willing to put in the effort to bridge the various infrastructure gaps, you could go right ahead, but we are not extending the already over-extended abstractions further for non-essential content additions.

In fact, now that I'm looking at these crafting requirements, I'm thinking we have the capabilities to extend these recipes to be a lot more reasonable, which is necessarily going to also make them less optimal.  Nine hours to make a finished firearm from scrap with hand tools just doesn't work, I expect that to be closer to two weeks to a month for a musket, much less a revolver.

> Each engine would need engine-specific components,

Well that's simple, engines are too complex for lone survivors to make anyway, that was easy. We are NOT going to model anything complex like that, so we don't have to worry about it.  Similarly, most other car parts either have light requirements (sheet metal or mild steel) or aren't feasible to make anyway (axles, springs, modern wheel assemblies, transmissions, etc).

> Like I said I didn't know we had essentially a new craft recipe freeze until fully realistic crafting became a thing. Is that how it's going to be all recipes now?

No, just ones that are trying to push the envelope.  You pushed too far, so I'm pushing back.  The default is "don't add it", which is what I told you when you first asked about this, you're the one that's not taking no for an answer.

The more extensive answer is I want barrel casting, barrel boring, multi-stage crafting, and massively extended craft times to consider any further extensions to firearm crafting, much less things like revolvers and rifling, which have their own requirements.

I'll be revisiting the existing firearm recipes and probably removing most of them as these criteria haven't been met in the past either.

**Pull request closed without merging**